File: Algorithm/Python/036. Valid Sudoku.py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-04-09 09:57:54
# @Last modified by:   LC
# @Last Modified time: 2016-04-10 16:22:11
# @Email: liangchaowu5@gmail.com


# 思路：字典存储行，列，块对应的数字，遍历一次即可判断
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
    
        col_dict = {}
        block_dict = {}
        for i in xrange(len(board)):
            row_dict = [] # just a row
            if i/3 not in block_dict:
                block_dict[i/3] = {}
            for j in xrange(len(board[i])):
                if j not in col_dict:
                    col_dict[j] = []
                if j/3 not in block_dict[i/3]:
                    block_dict[i/3][j/3] = []
                    
                num = board[i][j]
                if num == '.':
                    continue
                
                # check row
                if num not in row_dict:
                    row_dict.append(num)
                else:
                    return False
                
                # check column
                if num not in col_dict[j]:
                    col_dict[j].append(num)
                else:
                    return False
                
                # check block
                if num not in block_dict[i/3][j/3]:
                    block_dict[i/3][j/3].append(num)
                else:
                    return False
            
        return True
                    
                
                
                
        

File: Algorithm/Java/036. Valid Sudoku.java
/**
* Author: WuLC
* Date:   2016-12-03 19:44:43
* Last modified by:   WuLC
* Last Modified time: 2016-12-03 19:50:22
* Email: liangchaowu5@gmail.com
*/

// bit manipulation
public class Solution 
{
    public boolean isValidSudoku(char[][] board) 
    {
        int[] row = new int[9];
        int[] col = new int[9];
        int[][] block = new int[3][3];
        int num;
        for (int i = 0; i<9; i++)
        {
            for (int j=0; j<9; j++)
            {
                if (board[i][j] == '.') continue;
                if (board[i][j]>'9' || board[i][j]<'0') return false;
                num = board[i][j] - '0';
                // check row
                if ((row[i]&(1<<num)) != 0) return false;
                else row[i] ^= (1<<num);
                // check column
                if ((col[j]&(1<<num)) != 0) return false;
                else col[j] ^= (1<<num);
                // check block
                if ((block[i/3][j/3] & (1<<num)) != 0) return false;
                else block[i/3][j/3] ^= (1<<num);
                
                /* the above checking can be replaced by the followding code
                int tmp = 1<<num;
                if ( (row[i]&tmp)!=0 || (col[j]&tmp)!=0 || (block[i][j]&tmp)!= 0) return false;
                else
                {
                    row[i] ^= tmp;
                    col[j] ^= tmp;
                    block[i][j] ^= tmp;
                }
                 */
            }
        }
        return true;
    }
}